x-pack/plugins/security_solution/public/resolver/store/selectors.ts

```diff
@@ -122,3 +132,93 @@ function composeSelectors<OuterState, InnerState, ReturnValue>(
 ): (state: OuterState) => ReturnValue {
   return (state) => secondSelector(selector(state));
 }
+
+const spatiallyIndexedEntities = createSelector(
+  (state: ResolverState) => state,
+  function (state) {
+    const indexedProcessTree = indexedProcessTreeFactory(
+      dataSelectors.graphableProcesses(state.data)
+    );
+    const processNodeViewWidth = 360;
+    const processNodeViewHeight = 120;
+    const widths = dataSelectors.widthsOfProcessSubtrees(indexedProcessTree);
+
+    const positions = dataSelectors.processPositions(indexedProcessTree, widths);
+
+    const edgeLineSegments = dataSelectors.processEdgeLineSegments(
+      indexedProcessTree,
+      widths,
+      positions
+    );
+
+    const tree: rbush<IndexedEntity> = new rbush();
+    const processesToIndex: IndexedProcessNode[] = [];
+    const edgeLineSegmentsToIndex: IndexedEdgeLineSegment[] = [];
+    for (const [processEvent, position] of positions) {
+      const transformedPosition = applyMatrix3(position, dataSelectors.isometricTransformMatrix);
+      const [nodeX, nodeY] = transformedPosition;
+      const indexedEvent: IndexedProcessNode = {
+        minX: nodeX - 0.5 * processNodeViewWidth,
+        minY: nodeY - 0.5 * processNodeViewHeight,
+        maxX: nodeX + 0.5 * processNodeViewWidth,
+        maxY: nodeY + 0.5 * processNodeViewHeight,
+        position: transformedPosition,
+        entity: processEvent,
+        type: 'processNode',
+      };
+      processesToIndex.push(indexedEvent);
+    }
+    for (const edgeLineSegment of edgeLineSegments) {
+      const transformedSegment: Vector2[] = [];
+      for (const point of edgeLineSegment) {
+        transformedSegment.push(applyMatrix3(point, dataSelectors.isometricTransformMatrix));
+      }
+      const [[x1, y1], [x2, y2]] = transformedSegment;
+      const indexedLineSegment: IndexedEdgeLineSegment = {
+        minX: Math.min(x1, x2),
+        minY: Math.min(y1, y2),
+        maxX: Math.max(x1, x2),
+        maxY: Math.max(y1, y2),
+        entity: transformedSegment,
+        type: 'edgeLine',
+      };
+      edgeLineSegmentsToIndex.push(indexedLineSegment);
+    }
+    tree.load([...processesToIndex, ...edgeLineSegmentsToIndex]);
+    return tree;
+  }
+);
+
+const boundingBox = composeSelectors(cameraStateSelector, cameraSelectors.viewableBoundingBox);
+
+const currentBoundingBox = createSelector(
+  (state: ResolverState) => state,
+  function (state) {
+    return boundingBox(state)(Date.now());
+  }
+);
+
+export const visibleProcessNodePositionsAndEdgeLineSegments = createSelector(
+  spatiallyIndexedEntities,
```

Comment:
:grey_question: I think we talked about this before, but we might need to think about selecting in the "active" and "selected" nodes from the UI (they may or may not be the same) and also possibly the ones "adjacent" to them as designated by the processAdjacency (sp?) selector. The reason is, I'm not sure how it would play out to have those things point (via aria-flowto and other mechanics) to nodes that "don't exist".

Would it be possible/good to make this like `spatiallyIndexedAndActiveUIEntities`?

Even though this is kind of a big issue, I'm marking it with a :grey_question: because I don't think it should stand in the way of merging this PR (and we have an issue on the board to deal with a11y stuff for 7.9 that we could cover this under. I'd ask that if you decide _not_ to look into this, you add a comment to ( https://github.com/elastic/endpoint-app-team/issues/289 ) just so we know to track it.

Reply:
I'd note that in _most_ cases this shouldn't add any more nodes to render. The `active` one should always be centered and when you're zoomed out enough, at least the `up`,`left`, and `right` ones should already be visible. Sometimes the `selected` one might be outside the viewbox (after we implement more a11y fixes) so you could be looking at maybe a worst case of 5 (selected + 4 directions) + 5 (active + maybe 4 directions) == 10 extras.